.github/actions/update-extension-provider-images/update-images.py

```diff
@@ -0,0 +1,626 @@
+            continue
+    current_versions.sort()
+
+    if not current_versions:
+        existing_minor_tags = []
+        missing_minor_tags = []
+        for ver in available_versions:
+            tag = available_versions_map[ver]
+            if image_exists(oci_client, repository, tag):
+                existing_minor_tags.append(tag)
+            else:
+                missing_minor_tags.append(tag)
+        return {
+            "patch": [],
+            "minor": existing_minor_tags,
+            "missing": missing_minor_tags,
+        }
+
+    highest_current_ver = current_versions[-1]
+
+    highest_patch_for_minor = defaultdict(lambda: semver.Version(0))
+    for ver in current_versions:
+        key = (ver.major, ver.minor)
+        if ver > highest_patch_for_minor[key]:
+            highest_patch_for_minor[key] = ver
+
+    newer_patch_tags = []
+    newer_minor_tags = []
+    missing_tags = []
+
+    for ver in available_versions:
+        original_tag = available_versions_map[ver]
+        key = (ver.major, ver.minor)
+
+        if key in highest_patch_for_minor:
+            if ver > highest_patch_for_minor[key]:
+                if image_exists(oci_client, repository, original_tag):
+                    newer_patch_tags.append(original_tag)
+                else:
+                    missing_tags.append(original_tag)
+        elif ver > highest_current_ver:
+            if image_exists(oci_client, repository, original_tag):
+                newer_minor_tags.append(original_tag)
+            else:
+                missing_tags.append(original_tag)
+
+    return {
+        "patch": sorted(newer_patch_tags, key=parse_to_semver),
+        "minor": sorted(newer_minor_tags, key=parse_to_semver),
+        "missing": sorted(missing_tags, key=parse_to_semver),
+    }
+
+
+def update_images_data(images_data: Dict, new_versions_by_name: Dict[str, Dict]) -> List[Update]:
+    """
+    Update the in-memory images data structure with new versions and return structured update info.
+
+    Args:
+        images_data (Dict): The full data from images.yaml, loaded as a dictionary.
+        new_versions_by_name (Dict[str, Dict]): Maps image name to its new
+        'patch' and 'minor' versions.
+
+    Returns:
+        List[Update]: A list of structured objects detailing each update.
+    """
+    updates: List[Update] = []
+    images_by_name = defaultdict(list)
+    for image in images_data["images"]:
+        images_by_name[image["name"]].append(image)
+
+    all_new_entries_global = []
+
+    for name, image_list in images_by_name.items():
+        if name not in new_versions_by_name:
+            continue
+
+        new_entries_for_this_name = []
+        newer = new_versions_by_name[name]
+        has_target_version = any("targetVersion" in img for img in image_list)
+
+        if not has_target_version:
+            # Handle images without 'targetVersion' (singleton components)
+            all_newer_tags = newer["patch"] + newer["minor"]
```

Comment:
nit: as mentioned before, using semver-arithmetic, you can only tell if a version is _greater_ or _smaller_. You have no means to determine whether it is newer or older (in a chronological sense)